src/core/toga/widgets/base.py

```diff
@@ -86,15 +132,14 @@ def app(self):
 
     @app.setter
     def app(self, app):
-        if self._app is not None:
+        if not (self._app is None or app is None):
```

Comment:
The underlying logic behind this method is sufficient complex that it's probably worth adding some comments elaborating the exact conditions (in human-readable terms) that cause the two branches to trigger.

Reply:
Yes, that is a horrible line of code. I'll clean it up and add a comment.